### PR TITLE
Ngraph reference implementation: OneHot specifications review.

### PR DESCRIPTION
**Pull request by IvanNovoselov**
Updated specifications for the OneHot  operation. The most notable changes are:
- Input indices must be non-negative, since it is assumed in the ref and CPU plugin implementations.
- The depth must be positive (explicit check in the ngraph class).

### CODE REVIEW

**Comment by iefode** [comment]
@lazarevevgeny Could you please take a look?

File: docs/ops/sequence/OneHot_1.md [hunk 1]
```diff
@@ -32,14 +32,14 @@ Types of input scalars `on_value` and `off_value` should match and can be any of
...
+* **2**: `depth`: positive scalar (0D tensor) of any supported integer type that specifies the number of classes and thus the size of the one-hot dimension. Required.
+* **3**: `on_value`: scalar (0D tensor) of any supported type that fills the locations in output tensor specified in `indices`. Required.
+* **4**: `off_value`: scalar (0D tensor) of the same type as `on_value` that fills the locations not represented in `indices`. Required.
 
 **Outputs**:
 
-* **1** Output tensor of rank `N+1`, where `N` is a rank of input tensor `indices`. A new axis of the size `depth` is inserted at the dimension `axis`.
+* **1** A tensor of rank `N+1`, where `N` is a rank of the input tensor `indices`. A new axis of the size `depth` is inserted at the dimension `axis`. The output type is the same as the `on_value` type.
 
 **Examples**
```
Comment (lazarevevgeny) [inline docs/ops/sequence/OneHot_1.md:44]: Please, provide more complex example.

File: docs/ops/sequence/OneHot_1.md [hunk 2]
```diff
@@ -32,14 +32,14 @@ Types of input scalars `on_value` and `off_value` should match and can be any of
...
-* **4**: `off_value`: scalar (0D tensor) of any type that is the value that the locations not represented by indices in input take.
+* **1**: `indices`: input tensor with non-negative indices of any supported integer data type. Can be 0D. Required.
+* **2**: `depth`: positive scalar (0D tensor) of any supported integer type that specifies the number of classes and thus the size of the one-hot dimension. Required.
+* **3**: `on_value`: scalar (0D tensor) of any supported type that fills the locations in output tensor specified in `indices`. Required.
+* **4**: `off_value`: scalar (0D tensor) of the same type as `on_value` that fills the locations not represented in `indices`. Required.
 
 **Outputs**:
 
-* **1** Output tensor of rank `N+1`, where `N` is a rank of input tensor `indices`. A new axis of the size `depth` is inserted at the dimension `axis`.
+* **1** A tensor of rank `N+1`, where `N` is a rank of the input tensor `indices`. A new axis of the size `depth` is inserted at the dimension `axis`. The output type is the same as the `on_value` type.
```
Comment (lazarevevgeny) [inline docs/ops/sequence/OneHot_1.md:43]: Add section with supported data types and use them in the description of input/output tensors

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]
Most of the changes looks good, but miss section about supported data types.

**Comment by IvanNovoselov** [comment]
Just noticed that some [constant_folding tests](https://github.com/IvanNovoselov/openvino/blob/cbff7196577c0dd516b9a233ede45b358add238c/ngraph/test/constant_folding.cpp#L2422-L2426) do pass negative indices to OneHot. This seems a little bit strange, since all negative indices will be [statically casted to huge size_t numbers](https://github.com/IvanNovoselov/openvino/blob/cbff7196577c0dd516b9a233ede45b358add238c/ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp#L60-L65), and almost certainly ignored (at least for reasonable depth values, not smth like 2^30).  I’m not sure if this negative-indices feature is needed, but if it does, I suggest to add smth like “When some elements from the `indices` **are negative, or** greater or equal to the `depth` it is a well-formed operation…”.
@lazarevevgeny, what do you think?

File: docs/ops/sequence/OneHot_1.md (hunk 2 above)
Comment (IvanNovoselov) [reply docs/ops/sequence/OneHot_1.md:49]: The types section is added.

File: docs/ops/sequence/OneHot_1.md (hunk 1 above)
Comment (IvanNovoselov) [reply docs/ops/sequence/OneHot_1.md:50]: Changed the one-dimensional input example to a more complicated case, plus added a two-dim input example.

**Comment by lazarevevgeny** [comment]
> Just noticed that some [constant_folding tests](https://github.com/IvanNovoselov/openvino/blob/cbff7196577c0dd516b9a233ede45b358add238c/ngraph/test/constant_folding.cpp#L2422-L2426) do pass negative indices to OneHot. This seems a little bit strange, since all negative indices will be [statically casted to huge size_t numbers](https://github.com/IvanNovoselov/openvino/blob/cbff7196577c0dd516b9a233ede45b358add238c/ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp#L60-L65), and almost certainly ignored (at least for reasonable depth values, not smth like 2^30). I’m not sure if this negative-indices feature is needed, but if it does, I suggest to add smth like “When some elements from the `indices` **are negative, or** greater or equal to the `depth` it is a well-formed operation…”.
> @lazarevevgeny, what do you think?

CC @dmitry-gorokhov, @nshchego, @andrei-kochin One more example with negative indices like with Gather operation. 

@IvanNovoselov We preliminary agreed to that the behaviour with negative indices is undefined, because different plugins may work differently since we don't support negative indices.

**Comment by IvanNovoselov** [comment]
@lazarevevgeny could you please clarify what this undefined behavior means from the implementation point of view? Specifically, I would suggest:
1) The ref. implementation should not throw on negative indices, but have to ignore them silently (since they are not supported).
2) Negative indices should be eliminated from the constant folding tests for one_hot, because undefined behavior should not be tested.
3) This behavior should be explicitly stated in the specs, smth like "Behavior for negative indices is undefined."
Please, correct me if I'm getting smth wrong.

**Comment by lazarevevgeny** [comment]
> @lazarevevgeny could you please clarify what this undefined behavior means from the implementation point of view? Specifically, I would suggest:
> 
> 1. The ref. implementation should not throw on negative indices, but have to ignore them silently (since they are not supported).
> 2. Negative indices should be eliminated from the constant folding tests for one_hot, because undefined behavior should not be tested.
> 3. This behavior should be explicitly stated in the specs, smth like "Behavior for negative indices is undefined."
> 
> Please, correct me if I'm getting smth wrong.

Agree. Let's write 0 value to elements corresponding to negative elements.

**Review by IvanNovoselov — COMMENTED** [review]
All the suggested corrections were addressed.

**Comment by IvanNovoselov** [comment]
@lazarevevgeny, please take a look at the corrected version, and approve the changes if everything is Ok.

**Comment by lazarevevgeny** [comment]
Please, revert commit of the mkl-dnn.

File: docs/ops/sequence/OneHot_1.md [hunk 3]
```diff
@@ -32,35 +32,66 @@ Types of input scalars `on_value` and `off_value` should match and can be any of
 
 **Inputs**:
 
-* **1**: `indices`: input tensor of rank `N` with indices of any supported integer data type. Can be 0D. Required.
-* **2**: `depth`: scalar (0D tensor) of any supported integer type that specifies number of classes and the size of one-hot dimension.
-* **3**: `on_value`: scalar (0D tensor) of any type that is the value that the locations in output tensor represented by indices in input take.
-* **4**: `off_value`: scalar (0D tensor) of any type that is the value that the locations not represented by indices in input take.
+* **1**: `indices`: input tensor of type T1 with non-negative indices, behavior for negative indices is undefined. Can be 0D. Required.
```
Comment (lazarevevgeny) [inline docs/ops/sequence/OneHot_1.md:35]: Wrap types "T1" and "T2" with asterisk character like this: `...of type *T1* with...`

File: docs/ops/sequence/OneHot_1.md [hunk 4]
```diff
@@ -32,35 +32,66 @@ Types of input scalars `on_value` and `off_value` should match and can be any of
...
+* **4**: `off_value`: scalar (0D tensor) of type T2 that fills the locations not represented in `indices`. Required.
 
 **Outputs**:
 
-* **1** Output tensor of rank `N+1`, where `N` is a rank of input tensor `indices`. A new axis of the size `depth` is inserted at the dimension `axis`.
+* **1**: An `N+1` rank tensor of type T2, where `N` is a rank of the input tensor `indices`. A new axis of the size `depth` is inserted at the dimension `axis`.
+
+**Types**
+
+* *T1*: any supported integer data type.
```
Comment (lazarevevgeny) [inline docs/ops/sequence/OneHot_1.md:46]: Let's specify that we support `int32` and `int64` only.

**Review by lazarevevgeny — APPROVED** [review]
Two minor comments to the spec + major comment for the mkl-dnn. But approve in advance.

**Comment by ilyachur** [comment]
Looks like you changed the mkl-dnn submodule by mistake.

File: docs/ops/sequence/OneHot_1.md (hunk 3 above)
Comment (IvanNovoselov) [reply docs/ops/sequence/OneHot_1.md]: Corrected.

File: docs/ops/sequence/OneHot_1.md (hunk 4 above)
Comment (IvanNovoselov) [reply docs/ops/sequence/OneHot_1.md]: Ok, done.

**Comment by IvanNovoselov** [comment]
@ilyachur, you are right. Fixed.

**Pull request merged**